### DISCUSSION TITLE
How would I enable utime for the Robot Inventor Hub?

### DISCUSSION THREAD

**Discussion by MarkWard0110** [pybricks/support#1327] [Q&A] [answered]
With the Lego Robot Inventor hub's factory firmware, I could use utime ticks_us to benchmark code execution.  I want to do the same in PyBricks.  However, I get an error `no module named 'utime'` when I try to import it

I could build my own pybricks firmware with utime enabled.  I'm trying to figure it out, but I am over my head now.  I thought it could be done by setting `#define MICROPY_PY_UTIME_MP_HAL                 (1)`, but when I deployed the build, I got the same error as before.  I am seeking guidance to enable utime or resolve why I can't import the package in a script.

You guys have done a great job, considering I have figured out how to build and deploy PyBricks in about a day!

**Comment by dlech** [comment]
You will need to `#define MICROPY_PY_UTIME (1)` as well.

**Reply by MarkWard0110** [reply]
It didn't work for me.  I have added `#define MICROPY_PY_UTIME            (1)`  in `_common/mpconfigport.h` and I also modified `_common/micropython.c (pb_builtin_import)` error message to something slightly different to confirm it is my build on the device.  I get my customized error message but no luck loading the module.

I build and deploy using ` make -C bricks/primehub -j8 deploy` with the hub in DFU mode.

The python code I am sending to the hub is 

```
import utime

print("hello world")
print("ticks_us: ", utime.ticks_us())`
```
I run the script using `pybricksdev run ble  .\hellotime.py


I also tried importing other packages that were available on the Lego firmware like `math`, `sys`, and `random`.  Each error with no module named `math` and no module named `random`.  I wanted to confirm I could load a MicroPython module.  

Is there something different about PyBricks module loading?

Is there a better way to confirm if the build includes the expected modules?

**Reply by dlech** [reply]
> Is there something different about PyBricks module loading?

Yes, we use the `u` prefix on all MicroPython packages that are included, so `umath` instead of `math` and `urandom` instead of `random`, etc.

**Reply by MarkWard0110** [reply]
Yes, that worked for math and random!  I'll have to find where I could have learned that detail.   Maybe this information might help identify why the utime isn't found?   If PyBrikcs has customized the gap between MicroPython and user programs there is something else that needs to be done to bring utime into the system as a `uutime` ?

**Reply by dlech** [reply]
It would just be `utime`, not `uutime` since it already has the u prefix. You will also need to add `PY_EXTRA_SRC_C += ports/stm32/modutime.c` to `bricks/_common/arm_none_eabi.mk` (or make your own `modutime.c` based on that file).

**Reply by laurensvalk** [reply, marked as answer]
As a combination of this answer and the other answer below, you could add your own microsecond method to `pybricks/tools/pb_type_stopwatch.c` without adding any new files or build steps.

You could copy and modify `tools_StopWatch_time`, and have it return a variation of `mp_hal_ticks_us` instead of `mp_hal_ticks_ms`.

Also make sure to add an entry to `tools_StopWatch_locals_dict_table`, again similar to the `time` method.

**Reply by MarkWard0110** [reply]
Thank you @dlech and @laurensvalk.  I thought I just wanted to enable a python package instead I'm learning about time.  What I think I have figured out is.  mp_hal_ticks_* is abstract and implemented per ports (platform).  PyBricks overloads these for its StopWatch.  Many of the implementations when delaying ms will yield to other tasks.  The delaying us will spin on the CPU.  Many of the implementations track SysTicks with a ms time scale.  It is the Unix and Windows ports that have a base time scale of microseconds.

I have duplicated the StopWatch time function with microsecond, it worked in the VirtualHub.  I have not gotten a utime module in either PrimeHub or VirtualHub.  That may not matter as much now that I have discovered how these timers are working.

I want to mark both as answers!  @laurensvalk's answer was the easiest to get working.  It may not be a replacement for utime but it is a start at getting access to microseconds.  @dlech's answer helped me understand and figure out how MicroPython is working.  

There is a lot of information to unwrap now.  Especially how the choice of delay affects the system.

**Comment by laurensvalk** [comment]
> I could use utime ticks_us to benchmark code execution. I want to do the same in PyBricks. However, I get an error no module named 'utime' when I try to import it

Is it an option to repeat your code a few thousand times and test the duration with the `StopWatch` class with millisecond resolution?

When working with `ticks_us`, chances are your benchmarking code could be a significant factor by itself, and would be subject to more variation due to occasional garbage collection, which would be averaged out in the repeated test case.

**Reply by MarkWard0110** [reply]
I am in the process of porting my Lego Python to PyBricks.  I'm using utime, math, random from MicroPython.  While doing this, I am exploring how PyBricks and MicroPython work.  It isn't just benchmarking.  As a hobby, I'm exploring how time affects the system. I'm exploring what if the main loop did not govern the time resolution.  What if the time resolution was managed within the components and the loop didn't implicitly wait?   I'm still learning the math involved, and you guys may have a better grasp of time and how it affects the system.  I was investigating if I could add utime to help in my exploration. 

I know that utime overflows at 2^30 - 1, and `ticks_diff` can handle the overflow up to (2^30 - 1) / 2.  It was fun to learn and realize how time's data type affects the ability to have a simulation time.  Leading me to find out what I need and how this is done for real.  This is how I learned Python's numbers are not like other languages :) 

The perspective also depends on how the system is built.  I am exploring this with code that is like a game engine loop.  There is an update and render on each tick.  Each action in the update must calculate the new state and return.  Each render updates the motors and lights based on the state and return.  The update and render must avoid blocking calls.  Long-running tasks must be performed incrementally on each tick.    With a single thread, I have a responsive system that performs parallel actions.  The system optimizes the calls to update the lights and motors by performing it only once per tick. If the input for the motor has not changed, it skips calling the API, reducing the execution time.  Splitting the update and render helps reduce updating the light display to a single call.  No matter if one or more behaviors update the lights during the same tick.  The same is true for setting motor inputs.

**Reply by laurensvalk** [reply]
Awesome! Thanks for using Pybricks in your programs.

We do occasionally get requests to add microsecond resolution to the stopwatch, so maybe we should just do that as well at some point.
